server/keyspace_service.go

```diff
@@ -82,18 +85,32 @@ func (s *KeyspaceServer) WatchKeyspaces(request *keyspacepb.WatchKeyspacesReques
 	ctx, cancel := context.WithCancel(s.Context())
 	defer cancel()
 
-	err := s.sendAllKeyspaceMeta(ctx, stream)
+	revision, err := s.sendAllKeyspaceMeta(ctx, stream)
 	if err != nil {
 		return err
 	}
-	watchChan := s.client.Watch(ctx, path.Join(s.rootPath, endpoint.KeyspaceMetaPrefix()), clientv3.WithPrefix())
+
+	watcher := clientv3.NewWatcher(s.client)
+	defer watcher.Close()
+
 	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case res := <-watchChan:
-			keyspaces := make([]*keyspacepb.KeyspaceMeta, 0, len(res.Events))
-			for _, event := range res.Events {
+		rch := watcher.Watch(ctx, path.Join(s.rootPath, endpoint.KeyspaceMetaPrefix()), clientv3.WithPrefix(), clientv3.WithRev(revision))
+		for wresp := range rch {
+			if wresp.CompactRevision != 0 {
+				log.Warn("required revision has been compacted, use the compact revision",
+					zap.Int64("required-revision", revision),
+					zap.Int64("compact-revision", wresp.CompactRevision))
+				revision = wresp.CompactRevision
+				break
+			}
+			if wresp.Canceled {
+				log.Error("watcher is canceled with",
+					zap.Int64("revision", revision),
+					errs.ZapError(errs.ErrEtcdWatcherCancel, wresp.Err()))
+				return nil
```

Comment:
if wresp.Canceled is true, does it mean watcher.Watch() was cancelled or the watcher (got from clientv3.NewWatcher) was cancelled? If it's the former, can we call watcher.Watch() again?

Reply:
I think when the ctx is canceled, both of them are canceled.

Reply:
OK, if wresp.Canceled is true, we don't need to call watcher.Watch() because this is only case that watcher won't call  watcher.Watch() by itself. I checked https://github.com/Homebrew/brew/blob/4.0.10/bin/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/clientv3/watch.go.

For your case, you can just return error to the client and let the client to retry. For my case, keyspace group manager has to retry until etcd is recovered. Probablly I should call clientv3.NewWatcher(kgm.etcdClient) to create a new watcher.